[pt2, docs] Add new PT2 troubleshooting doc (#138620)
pytorch/pytorch: docs/source/torch.compiler_troubleshooting.rst

``torch.compile`` Troubleshooting

> =================================

Reviewer: So what's the plan for keeping this up to date with changes in gdoc

Author: This will become the source of truth - the troubleshooting gdoc is just a draft.

.. code-block:: py

    def bad1_inner(...):
        # skipped

    @torch.compiler.disable
    def bad1_outer(...):
        # skipped
        bad1_inner(...)

    def bad2_inner(...)
        # traced

    @torch.compiler.disable(recursive=False)
    def bad2_outer(...):
        # skipped
        bad2_inner(...)

    @torch.compile
    def fn(...):
        # graph break
        bad1_outer(...)
        ...
        # graph break
        bad2_outer(...)

For example, we use ``torch.compiler.disable`` to disable ``torch.compile`` on sparse architecture in
recommendation models, as the sparse arch is difficult to compile. Preprocessing and logging functions
are other examples of functions that typically cause a lot of graph breaks and do not get value from being compiled.

If you are experiencing compiler crashes and you want to YOLO keep going, you can set
``torch._dynamo.config.suppress_errors = True``. Whenever the compiler crashes, we will just skip tracing
the function and try again later. This is not best practice - it is better to eventually manually add
disable annotations as necessary.

Resolving graph breaks
----------------------

> Recall that to maximize optimization opportunities, we should reduce the number of graph breaks.

Reviewer 2: ```suggestion
To maximize optimization opportunities,  it's important to reduce the number of graph breaks.
```

> In general, graph breaks are caused by one of:

Reviewer 2: ```suggestion
In general, graph breaks are caused by one of the following:
```

- You're trying to do something that fundamentally cannot be traced, such as data-dependent control flow.

> - You're trying to do something that we haven't gotten around to supporting yet.

Reviewer 2: ```suggestion
- You're trying to do something not yet supported. .
```

  For example, we currently have limited support for tracing code that uses the built-in Python ``inspect`` module.
- Your code has an error in it. For example, you may have tried calling a function with an incorrect number of arguments.

Unfortunately, many graph breaks are not actionable without a deeper understanding of Dynamo.

> It can even be difficult to determine which of the 3 causes was the true cause of your graph break.

Reviewer 2: ```suggestion
It can even be challenging to determine which of the three causes was the true cause of your graph break.
```

        self.call_function(fn, args, kwargs)
    File "/data/users/williamwen/pytorch/torch/_dynamo/symbolic_convert.py", line 838, in call_function
        self.push(fn.call_function(self, args, kwargs))  # type: ignore[arg-type]
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/data/users/williamwen/pytorch/torch/_dynamo/variables/misc.py", line 1038, in call_function
        return self.obj.call_method(tx, self.name, args, kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/data/users/williamwen/pytorch/torch/_dynamo/variables/tensor.py", line 527, in call_method
        result = handler_method(*args, **kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/data/users/williamwen/pytorch/torch/_dynamo/variables/tensor.py", line 773, in method_item
        unimplemented("Tensor.item")
    File "/data/users/williamwen/pytorch/torch/_dynamo/exc.py", line 304, in unimplemented
        raise Unsupported(msg, case_name=case_name)
    torch._dynamo.exc.Unsupported: Tensor.item

The general workaround for these graph breaks is to avoid doing data-dependent operations. Some specific workarounds are:

> - If your control flow doesn't actually depend on data values, you can change your code to do control flow on constants

Reviewer 2: ```suggestion
- If your control flow doesn't actually depend on data values, consider modifying your code to perform control flow on constants.
```

    Fortunately, you can use ``TORCHDYNAMO_DEBUG_FUNCTION=<function name>``, which limits dynamo tracing to only functions with a specific name
    (exact match). This will allow you to filter all of the functions in the model to the function(s) of interest.
    Use this in combination with the above debugging strategies.

Bytecode generation errors
^^^^^^^^^^^^^^^^^^^^^^^^^^

It is possible (though not common) for Dynamo to generate incorrect bytecode. This might be the case if you determine:

- Ablation reveals the error is happening at the TorchDynamo level
- The error is not being emitted from TorchDynamo stack frames
- The error looks more like a user error rather than a Dynamo error, or is a segfault
- The error does not occur without ``torch.compile``

Bytecode generation bugs are generally tricky to fix and we recommend just submitting an issue.
If you are interested in seeing the bytecode that Dynamo generates, you can use ``TORCH_LOGS=bytecode``.
You can see a high-level overview on what bytecode Dynamo generates `here <https://docs.google.com/presentation/d/1tMZOoAoNKF32CAm1C-WfzdVVgoEvJ3lp/edit?usp=sharing&ouid=114922067987692817315&rtpof=true&sd=true>`__.

> AOTAutograd

Author: We don't apply this to AOTAutograd elsewhere in the document because it's a name, not code.

-----------

AOTAutograd errors are typically difficult to debug - we recommend just submitting an issue.
AOTAutograd logging output is primarily helpful to see what the input to Inductor is.

TorchInductor
-------------

.. TODO

.. _troubleshooting_torch_logs_options:

Summary of TORCH_LOGS options
---------------------------------

> A summary of helpful ``TORCH_LOGS`` options are:

Reviewer 2: ```suggestion
A summary of helpful ``TORCH_LOGS`` options is:
```

- **Context managers**. Did you replicate any active context managers?
  This could be ``torch.no_grad``, automatic mixed precision, ``TorchFunctionMode`` / ``TorchDispatchMode``,
  activation checkpointing, compiled autograd etc.
- **Tensor subclasses**. Is there a tensor subclass involved?

Minifier
--------

The minifier is an early ``torch.compile`` tool that, given an FX graph that crashes when we attempt to run or compile it,
finds a subgraph that also crashes and outputs the code that performs that subgraph's operations.
Essentially, the minifier finds a minimal repro for a certain class of ``torch.compile``-related crashes.
This assumes that we were able to successfully trace through code.

Unfortunately, most of the time nowadays, the minifier doesn't work as expected, and alternative methods may be necessary.
This is likely because bugs that can be automatically reproduced in this manner are generally easier to fix
and have already been addressed, leaving more complex issues that do not reproduce easily.
However, it is straightforward to attempt using the minifier, so it is worth trying even if it may not succeed.

> Instructions for operating the minifier can be found `here <https://pytorch.org/docs/stable/torch.compiler_troubleshooting_old.html>`__.

Reviewer 3: @williamwen42  This link seem to be dead. 404 in https://pytorch.org/docs/stable/torch.compiler_troubleshooting_old.html

Author: Try again? The link looks active to me (also on icognito).

Reviewer 3: > Try again? The link looks active to me (also on icognito).

oh, works now! Not sure what happened when I tried it earlier

Larger models may take upwards of 30 minutes to a few hours.

Terminology
~~~~~~~~~~~

The following terms are relevant to troubleshooting ``torch.compile`` problems.

Graph break
-----------

``torch.compile`` traces your code and attempts to capture your PyTorch code into a
single computation graph of PyTorch operators (FX graph). However, this is not always possible.
When encountering code that can't be traced, a "graph break" occurs.
A graph break involves compiling the FX graph has been determined so far, running the unsupported code,
then resuming tracing after the unsupported code with a new FX graph.
Because the computation graph is broken up, we lose optimization opportunities,
so model code should avoid graph breaks whenever possible.
Graph breaks occur on things like:

- Data-dependent if-statements
- Many Python built-in functions
- C functions

Below is an example of a graph break due to the function ``copy.deepcopy`` from a Python builtin library
(exact output may differ).

.. code-block:: py

    import copy
    import torch

    @torch.compile
    def fn(x):
        x = x + 1

>         y = copy.deepcopy(x)

Reviewer: This is not a great example, because in principle we can eventually support this builtin

Author: Switched to using `eval`

Reviewer: eval is a bad example too, eval is actually easy to support when we do bytecode interpretation LOL

Author: What about `open`?

Reviewer: that one's good!
